lib/SILOptimizer/FunctionSignatureTransforms/ExistentialTransform.cpp

```diff
@@ -0,0 +1,595 @@
+  SILFunctionConventions Conv(SubstCalleeType, M);
+  SILType ResultType = Conv.getSILResultType();
+
+  /// If the original function has error results,  we need to generate a
+  /// try_apply to call a function with an error result.
+  if (FunctionTy->hasErrorResult()) {
+    SILFunction *Thunk = ThunkBody->getParent();
+    SILBasicBlock *NormalBlock = Thunk->createBasicBlock();
+    ReturnValue =
+        NormalBlock->createPHIArgument(ResultType, ValueOwnershipKind::Owned);
+    SILBasicBlock *ErrorBlock = Thunk->createBasicBlock();
+
+    SILType Error = Conv.getSILType(FunctionTy->getErrorResult());
+    auto *ErrorArg =
+        ErrorBlock->createPHIArgument(Error, ValueOwnershipKind::Owned);
+    Builder.createTryApply(Loc, FRI, SubMap, ApplyArgs, NormalBlock,
+                           ErrorBlock);
+
+    Builder.setInsertionPoint(ErrorBlock);
+    Builder.createThrow(Loc, ErrorArg);
+    Builder.setInsertionPoint(NormalBlock);
+  } else {
+    /// Create the Apply with substitutions
+    ReturnValue = Builder.createApply(Loc, FRI, SubMap, ApplyArgs, false);
+  }
+
+  /// Set up the return results.
+  if (NewF->isNoReturnFunction()) {
+    Builder.createUnreachable(Loc);
+  } else {
+    Builder.createReturn(Loc, ReturnValue);
+  }
+}
+
+/// Strategy to specialize existentail arguments:
+/// (1) Create a protocol constrained generic function from the old function;
+/// (2) Create a thunk for the original function that invokes (1) including
+/// setting
+///     its inline strategy as always inline.
+void ExistentialTransform::createExistentialSpecializedFunction() {
+  std::string Name = createExistentialSpecializedFunctionName();
+  SILLinkage linkage = getSpecializedLinkage(F, F->getLinkage());
+
+  /// Create devirtualized function type.
+  auto NewFTy = createExistentialSpecializedFunctionType();
+
+  auto NewFGenericSig = NewFTy->getGenericSignature();
+  auto NewFGenericEnv = NewFGenericSig->createGenericEnvironment();
+
+  /// Step 1: Create the new protocol constrained generic function.
+  NewF = FunctionBuilder.createFunction(
+      linkage, Name, NewFTy, NewFGenericEnv, F->getLocation(), F->isBare(),
+      F->isTransparent(), F->isSerialized(), F->getEntryCount(), F->isThunk(),
+      F->getClassSubclassScope(), F->getInlineStrategy(), F->getEffectsKind(),
+      nullptr, F->getDebugScope());
+  /// Set the semantics attributes for the new function.
+  for (auto &Attr : F->getSemanticsAttrs())
+    NewF->addSemanticsAttr(Attr);
+
+  /// Set Unqualified ownership, if any.
+  if (!F->hasQualifiedOwnership()) {
+    NewF->setUnqualifiedOwnership();
+  }
+
+  /// Step 1a: Populate the body of NewF.
+  SubstitutionMap Subs =
+      SubstitutionMap::get(NewFGenericSig,
+                           [&](SubstitutableType *type) -> Type {
+                             return NewFGenericEnv->mapTypeIntoContext(type);
+                           },
+                           LookUpConformanceInSignature(*NewFGenericSig));
```

Comment:
Please just use `LookUpConformanceInModule` instead. `LookUpConformanceInSignature` is going away